torch/distributed/pipelining/stage.py

```diff
@@ -606,26 +613,42 @@ def backward_one_chunk(
                 "input_values": input_values,
             }
 
-        self.grads_input = self.backward_maybe_with_nosync(bwd_kwargs)
-        logger.debug(f"{self.log_prefix} Backwarded chunk {bwd_chunk_id}")  # noqa: G004
+        # self.grads_input = self.backward_maybe_with_nosync(bwd_kwargs)
+        # logger.debug(f"{self.log_prefix} Backwarded chunk {bwd_chunk_id}")  # noqa: G004
 
         if not full_backward:
-            assert self.dw_builder, "Must provide dw_builder to run partial backward"
-            assert bwd_chunk_id not in self.dw_runner, (
-                f"{self.log_prefix} Attempted to run partial backward for chunk {bwd_chunk_id}"
-                " repeatedly without calling `backward_weight_one_chunk`"
-            )
-            dw_runner = self.dw_builder()
-            self.dw_runner[bwd_chunk_id] = dw_runner
+            if not self.dw_builder:
+                logger.info("IN LOGGGSS")
+                # print(f"{bwd_kwargs['stage_output']}, {bwd_kwargs['input_values']}")
+                # if "stage_ouput" is a loss, then it is a tensor, otherwise it is a tuple of tensors
+                if isinstance(bwd_kwargs['stage_output'], torch.Tensor):
+                    bwd_kwargs['stage_output'] = (bwd_kwargs['stage_output'],)
+
+                # print(f"{bwd_kwargs['stage_output']}, {bwd_kwargs['input_values']}")
+                dinputs, param_groups = stage_backward_input(bwd_kwargs["stage_output"], bwd_kwargs["input_values"], self.submod.parameters())
+                self.backward_state[bwd_chunk_id] = (dinputs, input_values, param_groups)
```

Comment:
nit: maybe just create a dw_runner here (lambda that calls stage_backward_weight with the right param groups..) and push it onto the self.dw_runner queue. then in W execution you dont have to check which mode you're in.